docs/design/features/InlineArrayAttribute.md

```diff
@@ -0,0 +1,95 @@
+Our preexisting offering in this area – [unsafe fixed-sized buffers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#fixed-size-buffers) has multiple constraints, in particular it works only with blittable value types and provides no overrun/type safety, which considerably limits its use.
+
+*The InlineArrayAttribute is a building block to allow efficient, type-safe, overrun-safe indexable/sliceable inline data.*
+
+An example of one such feature is [`Inline Arrays` in C# 12.0](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/inline-arrays.md).
+
+### API
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Struct, AllowMultiple = false)]
+    public sealed class InlineArrayAttribute : Attribute
+    {
+        public InlineArrayAttribute (int length)
+        {
+            Length = length;
+        }
+
+        public int Length { get; }
+    }
+}
+```
+
+When `InlineArray` attribute is applied to a struct with one instance field, it is interpreted by the runtime as a directive to replicate the layout of the struct `Length` times. That includes replicating GC tracking information if the struct happens to contain managed pointers.
+
+Unlike "filed0; field1; field2;..." approach, the resulting layout would be guaranteed to have the same order and packing details as elements of an array with element `[0]` matching the location of the single specified field. That will allow the whole aggregate to be safely indexable/sliceable.
+
+`Length` must be greater than 0.
+
+struct must not have explicit layout.
+
+In cases when the attribute cannot have effect, it is an error case handled in the same way as the given platform handles cases when a type layout cannot be constructed.
+Generally, it would be a `TypeLoadException` thrown at the time of layout construction.
+
+### API Usage
+
+```C#
+// runtime replicates the layout of the struct 42 times 
+[InlineArray(Length = 42)] 
+struct MyArray<T> 
+{ 
+    private T _element0; 
+    public Span<T> SliceExample() 
+    { 
+        return MemoryMarshal.CreateSpan(ref _element0, 42); 
+    } 
+} 
+```
+### Memory layout of an inline array instance.
+
+The memory layout of a struct instance decorated with `InlineArray` attribute closely matches the layout of the element sequence of an array `T[]` with length == `Length`.
+In particular (using the `MyArray<T>` example defined above):
+* In unboxed form there is no object header or any other data before the first element.
+
+Example: assuming the instance is not GC-movable, the following holds: `(byte*)*inst == (byte*)inst._element0`
+
+* There is no additional padding between elements.
+
+Example: assuming the instance is not GC-movable and `Length > 1`, the following will yield a pointer to the second element: `(byte*)inst._element0 + sizeof(T)`
```

Comment:
The expression doesn't look like a valid C#.

Reply:
Right. Not sure why `&` are missing. 
I think I initially wanted to make examples in terms of `ref`, but since we do not formally allow math operations with `ref`, I'd need to use `Unsafe` and it felt like not helping to understand the layout invariants.